scripts/usb_scan.py

```diff
@@ -0,0 +1,618 @@
+                interface.get_node(), k, str(rule)))
+            if not hex_equal(rule[k], device[self._KEY_MAP_DEVICE[k]]):
+                return False
+
+        for k in [k for k in rule if k in self._KEY_MAP_INTERFACE]:
+            log_wrapper("check {} props[{}] against {}".format(
+                interface.get_node(), k, str(rule)))
+            if not hex_equal(rule[k], interface[self._KEY_MAP_INTERFACE[k]]):
+                return False
+
+        log_wrapper("found matching rule: " + str(rule))
+        return True
+
+    def check(self, device):
+        """check if a usb device can be passed through
+
+        if one of device's interfaces is denied by rule, the device is denied.
+        Normally, there should be a catch all rule at last line of policy file:
+            ALLOW: # Otherwise allow everything else
+        If there isn't, and no matching rule is found, the interface will also
+        be denied.
+
+        Only if all interfaces are allowed, the device is allowed to passed
+        through.
+
+        Example of a device before checking
+        UsbDevice: (u'1-2', {'idVendor': '17ef', 'ID_VENDOR_FROM_DATABASE':
+        u'Lenovo', 'bConfigurationValue': '1', 'bDeviceClass': '00',
+        'ID_MODEL_FROM_DATABASE': '', 'version': ' 2.00', 'idProduct': '383c',
+          'bNumInterfaces': ' 1', 'serial': 'H8RU2TGE', 'bcdDevice': '0100'})
+            |__ UsbInterface: (u'1-2:1.0', {'bInterfaceNumber': '00',
+            'bInterfaceClass': '08', 'bInterfaceProtocol': '50',
+            'bInterfaceSubClass': '06'})
+            |__ Allow: False
+
+        UsbDevice: (u'4-1.2', {'idVendor': '2717', 'ID_VENDOR_FROM_DATABASE':
+         '' , 'bConfigurationValue': '1', 'bDeviceClass': '00',
+        'ID_MODEL_FROM_DATABASE': '', 'version': ' 2.00', 'idProduct': 'ff48',
+          'bNumInterfaces': ' 2', 'serial': '7b06864', 'bcdDevice': '0404'})
+            |__ UsbInterface: (u'4-1.2:1.0', {'bInterfaceNumber': '00',
+            'bInterfaceClass': 'ff', 'bInterfaceProtocol': '00',
+            'bInterfaceSubClass': 'ff'})
+            |__ UsbInterface: (u'4-1.2:1.1', {'bInterfaceNumber': '01',
+            'bInterfaceClass': 'ff', 'bInterfaceProtocol': '01',
+            'bInterfaceSubClass': '42'})
+            |__ Allow: False
+
+        :param device:(UsbDevice) the device to check
+        :return:(bool) if allow pass through
+        """
+        log_wrapper("policy check: " + device.get_node())
+        for i in device.get_all_interfaces():
+            allow_interface = False
+            for r in self.rule_list:
+                if self.match_device_interface(r, device, i):
+                    if r[self._ALLOW]:
+                        log_wrapper("allow " + i.get_node())
+                        allow_interface = True
+                        break
+                    else:
+                        log_wrapper("deny " + i.get_node())
+                        return False
```

Comment:
Isn't it possilble that a rule further down re-allows the device? I think here we should set `allow_interface=False` and keep scanning.

Reply:
No.  according to https://support.citrix.com/article/CTX119722

The first match for any device is considered definitive. 
so it can't be re-allow.